source/common/grpc/async_client_manager_impl.h

```diff
@@ -50,18 +54,64 @@ class AsyncClientManagerImpl : public AsyncClientManager {
   AsyncClientManagerImpl(Upstream::ClusterManager& cm, ThreadLocal::Instance& tls,
                          TimeSource& time_source, Api::Api& api, const StatNames& stat_names);
 
+  RawAsyncClientSharedPtr
+  getOrCreateRawAsyncClient(const envoy::config::core::v3::GrpcService& config, Stats::Scope& scope,
+                            bool skip_cluster_check) override {
+    RawAsyncClientSharedPtr client;
+    client = raw_async_client_cache_->getCache(config);
+    if (client != nullptr) {
+      return client;
+    }
+    client = factoryForGrpcService(config, scope, skip_cluster_check)->create();
+    raw_async_client_cache_->setCache(config, client);
+    return client;
+  }
   // Grpc::AsyncClientManager
   AsyncClientFactoryPtr factoryForGrpcService(const envoy::config::core::v3::GrpcService& config,
                                               Stats::Scope& scope,
                                               bool skip_cluster_check) override;
 
 private:
+  class RawAsyncClientCache : public ThreadLocal::ThreadLocalObject {
+  public:
+    void setCache(const envoy::config::core::v3::GrpcService& config,
+                  const RawAsyncClientSharedPtr& client) {
+      std::uint64_t key = hashByType(config);
+      cache_[key] = client;
+    }
+    RawAsyncClientSharedPtr getCache(const envoy::config::core::v3::GrpcService& config) const {
+      std::uint64_t key = hashByType(config);
+      auto it = cache_.find(key);
+      if (it == cache_.end()) {
+        return nullptr;
+      }
+      return it->second;
+    }
+
+  private:
+    absl::flat_hash_map<uint64_t, RawAsyncClientSharedPtr> cache_;
```

Comment:
nit: put the cache_ below, with the other member vars.

Reply:
Thanks, will do.